clash-lib/src/Clash/Core/Evaluator/Semantics.hs

```diff
@@ -0,0 +1,372 @@
+  , evaluateLiteral
+  , evaluateDataWith
+  , evaluatePrimWith
+  , evaluateLam
+  , evaluateTyLam
+  , evaluateAppWith
+  , evaluateLetrecWith
+  , evaluateCaseWith
+  , evaluateCastWith
+  , evaluateTickWith
+  , quoteLiteral
+  , quoteDataWith
+  , quotePrimWith
+  , quoteLamWith
+  , quoteCastWith
+  , quoteTickWith
+  , quoteNeVar
+  , quoteNePrimWith
+  , quoteNeAppWith
+  , quoteNeTyAppWith
+  , quoteNeCaseWith
+  ) where
+
+import Control.Monad (foldM)
+import Control.Monad.State.Strict (State)
+import qualified Control.Monad.State.Strict as State
+import Data.Bifunctor (bimap)
+import Data.Bitraversable (bitraverse)
+import Data.Foldable (foldl')
+
+import BasicTypes (InlineSpec(..))
+
+import Clash.Core.DataCon
+import Clash.Core.Evaluator.Models
+import Clash.Core.Literal
+import Clash.Core.Term
+import Clash.Core.TermInfo
+import Clash.Core.TyCon
+import Clash.Core.Type
+import Clash.Core.Var
+import Clash.Driver.Types (Binding(..))
+
+isFullyApplied :: Type -> [Either Term Type] -> Bool
+isFullyApplied ty args =
+  length args == length (fst $ splitFunForallTy ty)
+
+etaExpand :: Term -> State Env Term
+etaExpand x =
+  case collectArgs x of
+    y@(Data dc, _) -> do
+      tcm <- State.gets envTcMap
+      expand tcm (dcType dc) y
+
+    y@(Prim p, _) -> do
+      tcm <- State.gets envTcMap
+      expand tcm (primType p) y
+
+    _ -> pure x
+ where
+  expand :: TyConMap -> Type -> (Term, [Either Term Type]) -> State Env Term
+  expand tcm ty (tm, args) = do
+    let missingArgTys = fst $ splitFunForallTy (applyTypeToArgs tm tcm ty args)
+    missingArgs <- traverse etaNameOf missingArgTys
+
+    pure $ mkAbstraction
+      (mkApps x (fmap (bimap Var VarTy) missingArgs))
+      missingArgs
+
+  etaNameOf :: Either TyVar Type -> State Env (Either Id TyVar)
+  etaNameOf = \case
+    Left tv  -> pure (Right tv)
+    Right ty -> Left <$> State.state (mkUniqueId "eta" ty)
+
+-- | Default implementation for looking up a variable in the environment.
+-- This checks both the local and global environments, and inlines global
+-- binders which are not marked as NOINLINE.
+--
+evaluateVarWith :: (Term -> State Env Value) -> Id -> State Env Value
+evaluateVarWith eval i = do
+  env <- State.get
+
+  if |  Just etv <- lookupLocal i env
+     -> either (State.withState (deleteLocal i) . eval) pure etv
+
+     |  Just b <- lookupGlobal i env
+     ,  canInline (bindingSpec b)
+     -> either (State.withState (deleteGlobal i) . eval) pure (bindingTerm b)
+
+     |  otherwise
+     -> pure (VNeu (NeVar i))
+ where
+  canInline spec = spec == Inline || spec == Inlinable
+
+-- | Default implementation for evaluating a literal.
+-- This simply wraps the literal up into a Value.
+--
+evaluateLiteral :: Literal -> State Env Value
+evaluateLiteral = pure . VLit
+
+-- | Default implementaion for evaluating data constructors. If the constructor
+-- is not fully applied, then it is eta-expanded and its eta-expanded form is
+-- evaluated to obtain a result.
```

Comment:
A nullary data constructor / primitive is fully applied, otherwise it isn't. I should probably update the documentation with a note to make clear these are kind of a degenerate case though (compared to the evaluation of data / primtives in `evaluateAppWith`

Reply:
> A nullary data constructor / primitive is fully applied, otherwise it isn't.

In hindsight this is a terrible sentence. What I meant to express is that if you go to evaluate an expression and it's simply a constructor / prim, then it's only fully applied if it's nullary. If not, it's eta expanded.